Fix flat output in uplc for deBruijn-indexed ASTs

kwxm opened the pull request:
I was trying to use `uplc` to convert some files to flat format and got an error like this:
```
$ cabal exec uplc convert -- -i plutus-benchmark/validation/data/auction_1-1.flat -o /dev/null --if flat --of flat
uplc: De-Bruijn-named ASTs are not supported for typed Plutus Core
```

The code that should handle this just raised an error for some reason.  This one-line PR gets it to do what it should (I think).  I'm not sure why it was like that in the first place (maybe my fault?).



<!--
IMPORTANT: if you are an external contributor, make sure you have read the "External contributors" section of CONTRIBUTING.

Here are some checklists you may like to use. Use your judgement.

This is just a checklist, all the normative suggestions are covered in more detail in CONTRIBUTING.
-->
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

kwxm commented on plutus-core/executables/Common.hs:
```diff
@@ -355,7 +355,7 @@ writeFlat ::
   flatProg <- case flatMode of
             Named         -> pure $ serialiseProgramFlat (() <$ prog) -- Change annotations to (): see Note [Annotation types].
-            DeBruijn      -> typedDeBruijnNotSupportedError -- TODO, this should work but we don't have a program here :/
+            DeBruijn      -> serialiseDbProgramFlat (() <$ prog)
```
Wait!  This works for untyped programs, but I think this code also handles typed ones and I'm not sure what'll happen then.

kwxm replied:
That seems to be OK.  The `uplc` command can convert the program below from PLC source to flat, but `pcl` fails with the "De-Bruijn-named ASTs are not supported for typed Plutus Core" message.

```
(program 1.1.1
[ (builtin multiplyInteger) (con integer 111) (con integer 222) ]
)
```

kwxm replied:
No, now I'm confused again, Aren't the `DeBruijn` and `NamedDeBruijn` cases doing exactly the same thing?

michaelpj replied:
There's some rather confusing ad-hoc polymorphism going on here, I'm quite confused about what is going on also.

kwxm replied:
I've worked out what's going on now and I think that `writeFlat` is doing the wrong thing.  We have at least three kinds of untyped ASTs, which use different types of Names:

  * `Named`: these have names containing both Uniques and textual identifiers.
  * `DeBruijn`: these have names which are just De Bruijn indices (no text), and are what we usually use for serialsation.
  * `NamedDeBruijn`: names are De Bruijn indices along with textual identifiers.
  
I'm not sure what `NamedDeBruijn` is for, despite the fact that it was probably me that added it (maybe for consistency with the CBOR stuff that we used to have).

Anyway, reading a program is handled by [`loadUplcASTfromFlat`](https://github.com/input-output-hk/plutus/blob/c1519ca8f65c8a87795d92b8c5a36a30047cbbd6/plutus-core/executables/Common.hs#L316).  It's given a parameter (ultimately coming from the `--if` option) which tells it what kind of thing to deserialise, and I think it does the right thing.  It reads in a flat file and converts it to Named format: for `Named` it just deserialises, for `NamedDeBruijn` it deserialises and applies `fromDeBruijn`, retaining the textual IDs, and for `DeBruijn` it deserialises, adds an "x" identifier to each name, and then applied `fromDeBruijn`.

However, writing a program is handled by `writeFlat` above.  For `Named` serialisation it calls the `serialiseProgramFlat` _function_, which does the right thing, but for `DeBruijn` (after my edit) and `NamedDeBruijn` it calls the `serialiseDbProgramFlat` _typeclass method_ for the `UplcProgram` type, which is implemented [here](https://github.com/input-output-hk/plutus/blob/c1519ca8f65c8a87795d92b8c5a36a30047cbbd6/plutus-core/executables/Common.hs#L96). That calls `toDeBruijn` (which throws away the textual IDs and then does the De Bruijn transformation) and then serialises, so it's always saving things in un-named De Bruijn indexed form, whereas for `NamedDeBruijn` we want to preserve the textual IDs.

This is all very confusing! I think there are at least three things we can do: 
  * Get rid of the `serialiseDbProgramFlat` method and use explicit functions instead.
  * Add an `AstNameType` parameter to `serialiseNamedDbProgramFlat`, which would make it consistent with the `loadASTfromFlat` method.
  * Do away with `NamedDeBruijn` serialisation altogether (and maybe even `Named` serialisation).
  
I'm not sure which of these is best. I'll think about it and try to remember why we have `Named` and `NamedDeBruijn` serialisation in the first place. Maybe these are useful for programs which analyse PLC scripts, since they contain more information than the simple De Bruijn indexed form.

kwxm replied:
Anyway, I think this PR solves my original problem.  However, `--of flatNameDeBruijn` does the wrong thing, as it did before.

thealmarty replied:
- when I refactored this I tried to keep the meaning of the code the same. I'm not sure if I succeeded. It may be the case that I didn't succeed.
- I do remember the not supported error though for `DeBruijn`. @kwxm do you remember maybe there was a reason we didn't want to support that?
- I like to get rid of things that we don't need. So I vote to do away with `NamedDeBruijn` if we're not using it. It can be in a follow up PR but let's fix things when we notice the problem, so we don't get confused again.

kwxm replied:
> I do remember the not supported error though for DeBruijn. @kwxm do you remember maybe there was a reason we didn't want to support that?

No, I don't remember.  I do want to support it though, since that's the format we use for transmission to the chain and it's useful to convert to and from that format.  I think I can cope without `NamedDeBruijn` though, although I'll think a bit more about that first. There's still a `flat` instance for that, so we can put it back if necessary.

I think it would make things a bit more consistent if we added  an `AstNameType` parameter to `serialiseNamedDbProgramFlat`, like `loadASTfromFlat`.  

If you're OK with those changes I could add them to this PR: I don't think it'd be too difficult.

kwxm replied:
Opps, just seen @michaelpj's reply.  In that case I vote to keep `NamedDeBruijn`, but we should add the `AstNameType` parameter to the serialisation method to make it easy to output that.

michaelpj replied:
>  In that case I vote to keep NamedDeBruijn, but we should add the AstNameType parameter to the serialisation method to make it easy to output that.

I've been wondering whether we should do something like what `cardano-api` does, where it has a JSON "envelope" format, which is just a two-field object where the first field tells you what "type" of thing it is, while the second field contains the serailized data. Then we could have typed input formats, allowing the commands to largely just DTRT, without having to put more tags into the "real" serialization format...

kwxm replied:
> a JSON "envelope" format

But wouldn't the flat-encoded stuff have to be encoded in some textual form then, making it significantly bigger?

michaelpj commented:
`NamedDeBruijn` is useful: e.g. on Nikos' branch it's what we use to evaluate terms using indices, but still preserve names so that if we get an error we get a more sensical result out.

kwxm commented:
I've renamed the ` serialiseDbProgramFlat` method in `Executable` to `serialiseProgramFlat` and added an `AstNameType` parameter to tell it how to do the serialisation. Another way to do this would have been to have separate instances of `Executable` for programs with different name types, but that would need a lot of instances only differing in one method.  I tested this manually and it appears to work for all three name formats.  Perhaps we should have automated tests to make sure that conversion between the different formats are correct, but I was too lazy to do that because I thought this was only going to be a one-line PR when I opened it.

kwxm commented on plutus-core/executables/Common.hs:
```diff
@@ -59,41 +59,49 @@ type PlcProg =
   UPLC.Program PLC.Name PLC.DefaultUni PLC.DefaultFun
 
-class Executable a where
+class Executable p where
```
I changed `a` to `p` here because we often use `a` for the annotation type and that was confusing me.

thealmarty commented on plutus-core/executables/Common.hs:
```diff
@@ -344,19 +359,12 @@ getProgram fmt inp =
 
 writeFlat ::
-  (Executable a, Functor a, Flat (a ())) => Output -> AstNameType -> a b -> IO ()
+  (Executable a, Functor a) => Output -> AstNameType -> a b -> IO ()
```
I guess these are in a lot of places. Only change if it bothers you. :)

kwxm replied:
I changed them all for consistency but now it looks strange because you don't usually see type parameters called p.

thealmarty reviewed (approved):
Thanks!

I will probably remove `ann` when I tidy up the parser.
